File: nodejs/stream_writer_transactional.ts
// Copyright Pravega Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {
    StreamTransactionGetTxnId,
    StreamTransactionIsOpen,
    StreamTransactionWriteEventBytes,
    StreamTransactionCommitTimestamp,
    StreamTransactionAbort,
    StreamTransactionToString,
    StreamTxnWriterBeginTxn,
    StreamTxnWriterGetTxn,
    StreamTxnWriterToString,
} from './native_esm.js';

/**
 * This represents a transaction on a given Stream.
 */
export interface Transaction {
    /**
     * Return the transaction id in unsigned integer 128 format.
     */
    get_txn_id: () => BigInt;

    /**
     * Check if the transaction is an OPEN state.
     */
    is_open: () => Promise<boolean>;

    /**
     * Write an event into the Pravega Stream.
     * The event of type string is converted into bytes with `UTF-8` encoding.
     *
     */
    write_event: (event: string, routing_key?: string) => Promise<void>;

    /**
     * Write a byte array into the Pravega Stream. This is similar to `write_event(...)`
     * api except that the the event to be written is a byte array.
     */
    write_event_bytes: (buf: Uint8Array, routing_key?: string) => Promise<void>;

    /**
     * Commit the Transaction.
     * This Causes all messages previously written to the transaction to go into the stream contiguously.
     * This operation will either fully succeed making all events consumable or fully fail such that none of them are.
     * There may be some time delay before readers see the events after this call has returned.
     */
    commit: () => Promise<void>;

    /**
     * Commit the Transaction and the associated timestamp.
     * This Causes all messages previously written to the transaction to go into the stream contiguously.
     * This operation will either fully succeed making all events consumable or fully fail such that none of them are.
     * There may be some time delay before readers see the events after this call has returned.
     */
    commit_timestamp: (timestamp: number) => Promise<void>;

    /**
     * Abort the Transaction.
     * Drops the transaction, causing all events written to it to be deleted.
     */
    abort: () => Promise<void>;

    /**
     * Returns the string representation of the transaction including its id and scoped stream.
     */
    toString: () => string;
}

/**
 * Returns a wrapped StreamWriter that helps users to call Rust code.
 *
 * Note: A Transaction cannot be created directly without using the StreamTxnWriter.
 */
export const Transaction = (transaction): Transaction => {
    const enc = new TextEncoder(); // A `string` to `Uint8Array` serializer.

    const get_txn_id = (): BigInt => BigInt(StreamTransactionGetTxnId.call(transaction));
    const is_open = (): Promise<boolean> => StreamTransactionIsOpen.call(transaction);
    const write_event = (event: string, routing_key?: string): Promise<void> =>
        StreamTransactionWriteEventBytes.call(transaction, enc.encode(event), routing_key);
    const write_event_bytes = (buf: Uint8Array, routing_key?: string): Promise<void> =>
        StreamTransactionWriteEventBytes.call(transaction, buf, routing_key);
    const commit = (): Promise<void> => StreamTransactionCommitTimestamp.call(transaction, Number.MIN_SAFE_INTEGER);
    const commit_timestamp = (timestamp: number): Promise<void> =>
        StreamTransactionCommitTimestamp.call(transaction, timestamp);
    const abort = (): Promise<void> => StreamTransactionAbort.call(transaction);
    const toString = (): string => StreamTransactionToString.call(transaction);

    return { get_txn_id, is_open, write_event, write_event_bytes, commit, commit_timestamp, abort, toString };
};

/**
 * This represents a Transaction writer for a given Stream.
 */
export interface StreamTxnWriter {
    /**
     * Create a new transaction.
     * This returns a StreamTransaction which can be perform writes on the created transaction.
     * It can also be used to perform commit() and abort() operations on the created transaction.
     */
    begin_txn: () => Promise<Transaction>;

    /**
     * Get a StreamTransaction for a given transaction id.
     */
    get_txn: (txn_id: BigInt) => Promise<Transaction>;

    /**
     * Returns the string representation the writer with its scoped stream.
     */
    toString: () => string;
}

/**
 * Returns a wrapped StreamTxnWriter that helps users to call Rust code.
 *
 * Note: A StreamTxnWriter cannot be created directly without using the StreamManager.
 */
export const StreamTxnWriter = (stream_txn_writer): StreamTxnWriter => {
    const begin_txn = async (): Promise<Transaction> =>
        Transaction(await StreamTxnWriterBeginTxn.call(stream_txn_writer));
    const get_txn = async (txn_id: BigInt): Promise<Transaction> =>
        Transaction(await StreamTxnWriterGetTxn.call(stream_txn_writer, txn_id.toString()));
    const toString = (): string => StreamTxnWriterToString.call(stream_txn_writer);

    return { begin_txn, get_txn, toString };
};

File: nodejs/src/stream_writer_transactional.rs
// Copyright Pravega Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use neon::prelude::*;
use pravega_client::event::transactional_writer::TransactionalEventWriter;
use pravega_client_shared::ScopedStream;
use pravega_client_shared::TxId;
use std::sync::Arc;
use tokio::runtime::Handle;
use tokio::sync::Mutex;

use crate::transaction::StreamTransaction;

impl Finalize for StreamTxnWriter {}

///
/// This represents a Stream writer for a given Stream.
/// Note: A StreamTxnWriter cannot be created directly without using the StreamManager.
///
pub(crate) struct StreamTxnWriter {
    writer: Arc<Mutex<TransactionalEventWriter>>,
    runtime_handle: Handle,
    stream: ScopedStream,
}

impl StreamTxnWriter {
    pub fn new(writer: TransactionalEventWriter, runtime_handle: Handle, stream: ScopedStream) -> Self {
        StreamTxnWriter {
            writer: Arc::new(Mutex::new(writer)),
            runtime_handle,
            stream,
        }
    }

    ///
    /// Create a new transaction.
    /// This returns a StreamTransaction which can be perform writes on the created transaction. It
    /// can also be used to perform commit() and abort() operations on the created transaction.
    ///
    pub fn js_begin_txn(mut cx: FunctionContext) -> JsResult<JsPromise> {
        let stream_txn_writer = cx
            .this()
            .downcast_or_throw::<JsBox<StreamTxnWriter>, _>(&mut cx)?;

        let channel = cx.channel();
        let (deferred, promise) = cx.promise();

        // spawn an `async` task on the tokio runtime.
        let writer = Arc::clone(&stream_txn_writer.writer);
        let runtime_handle = stream_txn_writer.runtime_handle.clone();
        stream_txn_writer.runtime_handle.spawn(async move {
            // run an async function in the tokio thread
            let result = writer.lock().await.begin().await;

            // notify and execute in the javascript main thread
            deferred.settle_with(&channel, move |mut cx| match result {
                Ok(txn) => Ok(cx.boxed(StreamTransaction::new(txn, runtime_handle))),
                Err(e) => cx.throw_error(e.to_string()),
            })
        });

        Ok(promise)
    }

    ///
    /// Get a StreamTransaction for a given transaction id.
    ///
    pub fn js_get_txn(mut cx: FunctionContext) -> JsResult<JsPromise> {
        let stream_txn_writer = cx
            .this()
            .downcast_or_throw::<JsBox<StreamTxnWriter>, _>(&mut cx)?;
        // convert a string back to u128, js should verify the txn_id
        // to be string containing only number
        let txn_id = cx
            .argument::<JsString>(0)?
            .value(&mut cx)
            .parse::<u128>()
            .unwrap();

        let channel = cx.channel();
        let (deferred, promise) = cx.promise();

        // spawn an `async` task on the tokio runtime.
        let writer = Arc::clone(&stream_txn_writer.writer);
        let runtime_handle = stream_txn_writer.runtime_handle.clone();
        stream_txn_writer.runtime_handle.spawn(async move {
            // run an async function in the tokio thread
            let result = writer.lock().await.get_txn(TxId(txn_id)).await;

            // notify and execute in the javascript main thread
            deferred.settle_with(&channel, move |mut cx| match result {
                Ok(txn) => Ok(cx.boxed(StreamTransaction::new(txn, runtime_handle))),
                Err(e) => cx.throw_error(e.to_string()),
            })
        });

        Ok(promise)
    }

    ///
    /// Return the string representation.
    ///
    pub fn js_to_str(mut cx: FunctionContext) -> JsResult<JsString> {
        let stream_txn_writer = cx
            .this()
            .downcast_or_throw::<JsBox<StreamTxnWriter>, _>(&mut cx)?;

        Ok(cx.string(format!("Stream: {:?} ", stream_txn_writer.stream)))
    }
}
